hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -919,4 +886,42 @@ private Map<String, HoodieRecord<HoodieMetadataPayload>> fetchBaseFileAllRecords
       return SecondaryIndexKeyUtils.getRecordKeyFromSecondaryIndexKey(record.getRecordKey());
     }, record -> record));
   }
+
+  @VisibleForTesting
+  public static Map<String, String> reverseLookupSecondaryKeysInternal(List<String> recordKeys,
+                                                                       Map<String, HoodieRecord<HoodieMetadataPayload>> baseFileRecords,
+                                                                       HoodieMetadataLogRecordReader logRecordScanner) {
+    Map<String, String> recordKeyMap = new HashMap<>();
+    Set<String> keySet = new TreeSet<>(recordKeys);
+    Set<String> deletedRecordsFromLogs = new HashSet<>();
+    Map<String, HoodieRecord<HoodieMetadataPayload>> logRecordsMap = new HashMap<>();
+    // Note that: we read the log records from the oldest to the latest!!!
+    // If we change the read order, we need update the following logic accordingly.
+    logRecordScanner.getRecords().forEach(record -> {
+      String recordKey = SecondaryIndexKeyUtils.getRecordKeyFromSecondaryIndexKey(record.getRecordKey());
+      HoodieMetadataPayload payload = record.getData();
+      if (!payload.isDeleted()) { // process only valid records.
+        if (keySet.contains(recordKey)) {
+          logRecordsMap.put(recordKey, record);
```

Comment:
should we also do 
```
deletedRecordsFromLogs.remove(recordKey);
```
within the if block. 
just incase a record is deleted first and added back later. we should not treat it as deleted.

Reply:
No, we cannot. The single usage of `deletedRecordsFromLogs` is to filter records from base file, .e.g., we have `br`, `lr1`(deleted), `lr2`(non-deleted). When we handle `lr2`, we cannot remove its `rk`, since `rk` is used to filter `br`.

Reply:
do we have test for this scenario ? 
i.e. 

add a record to logfile1, delete in logfile2 and then add it back in logfile3. 
if the test works, I am good.